### DISCUSSION TITLE
Proposal: "self" keyword for generic types

### DISCUSSION THREAD

**Discussion by pelletburner** [dotnet/csharplang#6702] [Language Ideas] [answered]
I would like to see a "self" keyword with the following semantics:

1. Can be used in generic members (such as "`ILogger<self> _logger`", to have an example) and generic invocations (example: "`loggerFactory.CreateLogger<self>()`"
2. Will always equal the exact type of the class that is containing the member / being currently executed

Clarification:

Imagine "`protected readonly ILogger _logger;`" in the abstract base class "`BaseClass`" and having the "`CreateLogger`" code from above in the constructor.

When implementing a new class "`Calculator`" based on "`BaseClass`" the logger that is created is automatically of type "`ILogger<Calculator>`" (NOT "`ILogger<BaseClass>`").

Reasoning:

Would like to have more polymorphy and write much less non-sensical boilerplate code just to have things like loggers with a proper logging context.

**Comment by yaakov-h** [comment]
I think there might be some angled brackets that are getting lost in your post. Can you please surround the code examples with backticks so that they get rendered as code?

**Reply by pelletburner** [reply]
Done. Sorry, I'm new here.

**Comment by Joe4evr** [comment, marked as answer]
#5413

Beyond that, I also have an attribute+analyzer for this in [JiiLib.Constraints](https://github.com/Joe4evr/JiiLib/tree/master/src/JiiLib.Constraints), tho be noted that testing on it has been limited.
